### ISSUE TITLE
CA1710 enhancement: Allow more suffixes for ICollection

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/roslyn-analyzers#4513]
### Analyzer

**Diagnostic ID**: [CA1710](https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1710)

### Describe the improvement

From the documentation, implementing for example `System.Collections.ICollection` requires having "Collection" suffix. I think it's reasonable to allow other suffixes as well (`Dictionary`, `Set`, etc).

The rule also doesn't enforce a suffix for implementing [ISet](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.iset-1?view=net-5.0) interface. It should be considered.

**Comment by Youssef1313** [comment]
This needs a design review.

**Comment by Evangelink** [comment]
It seems to be already available when looking at the code and doc:

![image](https://user-images.githubusercontent.com/11340282/101328912-f0a88280-3870-11eb-97b2-49ad76d395bc.png)

**Comment by Youssef1313** [comment]
> It seems to be already available when looking at the code and doc:

I haven't looked at the code, but the doc says it allows only "Collection" suffix for ICollection interface.
The doc also doesn't mention ISet interface nor IReadOnlyCollection

**Comment by Evangelink** [comment]
I am fine with tweaking the rule to add:

1. `ISet` gives suffix `Set`
2. `IReadOnlyCollection` gives suffix `Collection`

but if go forward it means that the users will potentially have more issues raised so it should be considered as part of a major feature.

**Comment by mavasani** [comment]
Agreed with @Evangelink. Seems reasonable to add the new features, but please target the PR to https://github.com/dotnet/roslyn-analyzers/tree/release/6.0.1xx-preview1 as it is essentially adding new diagnostics.

**Comment by Youssef1313** [comment]
@mavasani Should the changes be limited to the two cases @Evangelink mentioned, but not only ICollection to end with for example "Dictionary" (or other collection types)

**Comment by mavasani** [comment]
@Youssef1313 - sorry I didn't quite get your question. Can you give a specific example?

**Comment by Youssef1313** [comment]
For something like the following, the rule requires the class to end with "Collection" but doesn't allow it to end with "Dictionary", "Set", "List", "Stack", "Queue", etc.

```csharp
public class MyDictionary<T> : ICollection<T>
{
    // .......
}
```

A similar case in this repo is:

https://github.com/dotnet/roslyn-analyzers/blob/525d43d2fd640c4f8d601c1d0090bd40789be9d7/src/Utilities/Compiler/PooledObjects/PooledConcurrentSet.cs#L16-L19

**Comment by mavasani** [comment]
@Youssef1313 - Ah, yes. That seems like a reasonable change to me.

**Comment by Evangelink** [comment]
I don't know for the second part. This `PooledConcurrentSet` should instead implement `ISet`. If we still want to go in this direction then it should be uniform (i.e. allow all suffixes for `ICollection`, `ICollection<T>`, `IEnumerable`, `IEnumerable<T>` and `IReadOnlyCollection<T>`).

**Comment by Evangelink** [comment]
@mavasani @Youssef1313 I will do a PR to fix the `IReadOnlyCollection` and `ISet`, if you still want to go on with allowing the various suffixes when implementing collection and enumerable, please create a separate ticket.
As said previously, I am not really a big fan of this. If you want to implement a set then you should implement `ISet` or accept the suppression.

**Comment by sharwell** [comment]
It seems reasonable to _allow_ the `Set` suffix for `IEnumerable<T>` even though it would not be the _preferred_ suffix. If `PooledConcurrentSet` doesn't need (or use) the complexity of the members added by `ISet<T>`, it shouldn't be required to implement them. The result is still a set-like type.

If the type implements `ISet<T>` or [`IReadOnlySet<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlyset-1?view=net-5.0) then `Set` would be the preferred suffix.

**Comment by Evangelink** [comment]
Looking back at the implementation we do allow the `Collection` suffix for `Queue`, `Stack` and other specialized collections so I am fine with allowing specialized suffix for generic collections. I will update the PR.
